packages/core/contracts/financial-templates/long-short-pair/LongShortPair.sol

```diff
@@ -269,66 +267,143 @@ contract LongShortPair is Testable, Lockable {
-        // Create an instance of the oracle and get the price. If the price is not resolved revert.
-        OptimisticOracleInterface optimisticOracle = _getOptimisticOracle();
-        require(optimisticOracle.hasPrice(address(this), priceIdentifier, requestedTime, customAncillaryData));
-        int256 oraclePrice = optimisticOracle.settleAndGetPrice(priceIdentifier, requestedTime, customAncillaryData);
-
-        return oraclePrice;
+    // Return the oracle price for a given request timestamp and ancillary data combo.
+    function _getOraclePrice(uint64 requestTimestamp, bytes memory requestAncillaryData) internal returns (int256) {
+        return _getOptimisticOracle().settleAndGetPrice(priceIdentifier, requestTimestamp, requestAncillaryData);
     }
 
-    function _requestOraclePriceExpiration() internal {
+    // Request a price in the optimistic oracle for a given request timestamp and ancillary data combo. Set the bonds
+    // accordingly to the deployer's parameters. Will revert if re-requesting for a previously requested combo.
+    function _requestOraclePrice(uint256 requestTimestamp, bytes memory requestAncillaryData) internal {
         OptimisticOracleInterface optimisticOracle = _getOptimisticOracle();
 
         // Use the prepaidProposerReward as the proposer reward.
         if (prepaidProposerReward > 0) collateralToken.safeApprove(address(optimisticOracle), prepaidProposerReward);
         optimisticOracle.requestPrice(
             priceIdentifier,
-            expirationTimestamp,
-            customAncillaryData,
+            requestTimestamp,
+            requestAncillaryData,
             collateralToken,
             prepaidProposerReward
         );
 
         // Set the Optimistic oracle liveness for the price request.
         optimisticOracle.setCustomLiveness(
             priceIdentifier,
-            expirationTimestamp,
-            customAncillaryData,
+            requestTimestamp,
+            requestAncillaryData,
             optimisticOracleLivenessTime
         );
 
         // Set the Optimistic oracle proposer bond for the price request.
-        optimisticOracle.setBond(
-            priceIdentifier,
-            expirationTimestamp,
-            customAncillaryData,
-            optimisticOracleProposerBond
+        optimisticOracle.setBond(priceIdentifier, requestTimestamp, requestAncillaryData, optimisticOracleProposerBond);
+    }
+
+    // Fetch the optimistic oracle expiration price. If the timestamp is before expiration tries to fetch the
+    // early expiration timestamp and ancillary data. Else, fetch the price for the provided expiration timestamp and
+    // ancillary data. If the time is before the expiration timestamp (i.e early expiration) and the price is the ignore
+    // early expiration price then revert. This acts to block settlement on bad early expiration attempts.
+    function getExpirationPrice() internal {
+        if (getCurrentTime() < expirationTimestamp) {
```

Comment:
This condition strikes me a bit odd. It doesn't seem like this perfectly sorts between early and regular expiry (unless I'm missing something).

Couldn't the current time be after the expiration timestamp _even if_ we expired it early?

Reply:
Can we not check the early expiration timestamp or something? Is there no state that tells us which one has been triggered so this flow is simplified? Could we take advantage of the normal expire function to wipe away any early expiry state that might be hanging around?

Reply:
I've drastically simplified this. PTAL